backend/src/main/scala/bloop/io/ClasspathHasher.scala

```diff
@@ -66,219 +49,103 @@ object ClasspathHasher {
+                  FileHash.of(path, ByteHasher.hashFileContents(path.toFile))
+                cacheMetadataJar.put(path, (currentMetadata, newHash))
+                newHash
               }
-            )
           }
         }
       }
 
-      val initEntries = Task {
-        classpath.zipWithIndex.foreach {
-          case (absoluteEntry, idx) =>
-            acquireHashingEntry(absoluteEntry.underlying, idx)
-        }
-      }.doOnCancel(Task { isCancelled.compareAndSet(false, true); () })
-
-      // Let's first turn the obtained hash tasks into an observable, don't allow cancellation
-      val acquiredTask = Observable.fromIterable(acquiredByThisHashingProcess)
-
-      val cancelableAcquiredTask = Task.create[Unit] { (scheduler, cb) =>
-        val (out, _) = parallelConsumer.createSubscriber(cb, scheduler)
-        val _ = acquiredTask.subscribe(out)
-        Cancelable { () =>
-          isCancelled.compareAndSet(false, true); ()
+      def hash(path: Path): Task[FileHash] = Task {
+        try {
+          if (cancelCompilation.isCompleted) {
+            BloopStamps.cancelledHash(path)
+          } else if (isCancelled.get) {
+            cancelCompilation.trySuccess(())
+            BloopStamps.cancelledHash(path)
+          } else {
+            val hash = hashFile()
+            isHashingComplete.set(true)
+            hash
+          }
+        } catch {
+          // Can happen when a file doesn't exist, for example
+          case NonFatal(_) => BloopStamps.emptyHash(path)
         }
       }
 
-      initEntries.flatMap { _ =>
-        cancelableAcquiredTask
-          .doOnCancel(Task { isCancelled.compareAndSet(false, true); () })
-          .flatMap { _ =>
-            if (isCancelled.get || cancelCompilation.isCompleted) {
-              cancelCompilation.trySuccess(())
-              Task.now(Left(()))
-            } else {
-              Task.sequence(acquiredByOtherTasks.toList).map { _ =>
-                val hasCancelledHash = classpathHashes.exists(_.hash() == BloopStamps.cancelledHash)
-                if (hasCancelledHash || isCancelled.get || cancelCompilation.isCompleted) {
-                  cancelCompilation.trySuccess(())
-                  Left(())
-                } else {
-                  Right(classpathHashes.toVector)
-                }
-              }
-            }
-          }
+      /*
+       * As a protective measure, set up a timeout for hashing which complete hashing with empty value to unlock
+       * downstream clients on the assumption that the hashing of this
+       * entry is too slow because of something that happened to this
+       * process. The completion of the downstream promise will also log a
+       * warning to the downstream users so that they know that a hashing
+       * process is unusually slow.
+       *
+       * There is no need for specyfying `doOnCancel` callback, this task won't start before timeout (see timeoutTo implementation)
+       */
```

Comment:
> Is there a case for this measure?

Good question, my answer to this is.. I have no idea :)

> btw, what will happen if we return emptyHash instead of correct one?

Another good question without answer :)

I can only say that I didn't want to change too much, I just wanted to fix NPE, uncomment tests and simplify this mechanism. As for me, we can remove this timeout fallback and see what will happen. In cases like this, correctness shouldn't be put before throughput.

Reply:
Ouch. I missed that this timeout already was there. 
I incorrectly decided that you added that.

Ok, let's keep it as is.